nixos/modules/system/boot/loader/raspberrypi/uboot-builder.sh

```diff
@@ -21,18 +22,35 @@ copyForced() {
 
 # Add the firmware files
 fwdir=@firmware@/share/raspberrypi/boot/
+copyForced $fwdir/overlays  $target/overlays
 copyForced $fwdir/bootcode.bin  $target/bootcode.bin
 copyForced $fwdir/fixup.dat     $target/fixup.dat
+copyForced $fwdir/fixup4.dat    $target/fixup4.dat
+copyForced $fwdir/fixup4cd.dat  $target/fixup4cd.dat
+copyForced $fwdir/fixup4db.dat  $target/fixup4db.dat
+copyForced $fwdir/fixup4x.dat   $target/fixup4x.dat
 copyForced $fwdir/fixup_cd.dat  $target/fixup_cd.dat
 copyForced $fwdir/fixup_db.dat  $target/fixup_db.dat
 copyForced $fwdir/fixup_x.dat   $target/fixup_x.dat
 copyForced $fwdir/start.elf     $target/start.elf
+copyForced $fwdir/start4.elf    $target/start4.elf
+copyForced $fwdir/start4cd.elf  $target/start4cd.elf
+copyForced $fwdir/start4db.elf  $target/start4db.elf
+copyForced $fwdir/start4x.elf   $target/start4x.elf
 copyForced $fwdir/start_cd.elf  $target/start_cd.elf
 copyForced $fwdir/start_db.elf  $target/start_db.elf
 copyForced $fwdir/start_x.elf   $target/start_x.elf
 
+# Add the config.txt
+copyForced @configTxt@ $target/config.txt
+
 # Add the uboot file
 copyForced @uboot@/u-boot.bin $target/u-boot-rpi.bin
 
-# Add the config.txt
-copyForced @configTxt@ $target/config.txt
+# Add the raspberry pi 4 specific files
+if [[ "$version" == "4" ]]; then
```

Comment:
Any reason to copy the dtb files for pi4 here, only if it's pi4, but all the previous pi4 specific files anyway?

Couldn't you produce an image working for both pi3 and pi4?

Reply:
1. No, I was considering that myself. For one, I don't know exactly which firmware files are applicable on each system, so this was kind of just a guess. But the *4* files could probably be made pi4-specific too. I'll try to at least make this more consistent than it is now...

2. While I do think there is an opportunity for this particular thing to become "uboot-multipi" instead of just for a single pi, that would:
  a) require more changes (we only place a single u-boot bin right now, unlike the SD card setup)
  b) potentially be breaking -> we'd have to change how config.txt is handled and I'd want to be careful that the Pi handles config.txt flexibly enough that the user supplied extra config composes with what we'd want to lay down

I have kind of imagined this "uboot-mutlipi" thing happening, then could be used to generate the SD Card image and then some other nebulous clean-up goals...